### PR TITLE
Drop 32-bit

### LINKED ISSUE #10675: 4.1rc2: test_compound_fitting_with_units fails on 32-bit platforms
### Description

The model in `test_compound_fitting_with_units` does not only produce different results between Linux and macOS (#10414, #10415) but also between Linux x86_64 and i586. The former passes all tests, the latter produces this failure:

```
[   96s] + /usr/bin/python3 -B -c 'import astropy, sys; sys.exit(astropy.test(args="-v "))'
[   99s] ============================= test session starts ==============================
[   99s] platform linux -- Python 3.8.5, pytest-4.6.9, py-1.9.0, pluggy-0.13.1 -- /usr/bin/python3
[  100s] cachedir: .pytest_cache
[  100s] 
[  100s] Running tests with Astropy version 4.1rc2.
[  100s] Running tests in /usr/lib/python3.8/site-packages/astropy.
[  100s] 
[  100s] Date: 2020-08-24T12:59:50
[  100s] 
[  100s] Platform: Linux-5.8.0-1-pae-i686-with-glibc2.1.3
[  100s] 
[  100s] Executable: /usr/bin/python3
[  100s] 
[  100s] Full Python Version: 
[  100s] 3.8.5 (default, Jul 20 2020, 17:41:41) [GCC]
[  100s] 
[  100s] encodings: sys: utf-8, locale: UTF-8, filesystem: utf-8
[  100s] byteorder: little
[  100s] float info: dig: 15, mant_dig: 15
[  100s] 
[  100s] Package versions: 
[  100s] Numpy: 1.19.1
[  100s] Scipy: 1.5.2
[  100s] Matplotlib: 3.3.0
[  100s] h5py: 2.10.0
[  100s] Pandas: 1.0.5
[  100s] PyERFA: 1.7.0
[  100s] Cython: 0.29.21
[  100s] Scikit-image: not available
[  100s] asdf: 2.6.0
[  100s] 
[  100s] Using Astropy options: remote_data: none.
[  100s] 
[  100s] hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/abuild/rpmbuild/BUILD/.hypothesis/examples')
[  100s] Matplotlib: 3.3.0
[  100s] Freetype: 2.6.1
[  100s] rootdir: /usr/lib/python3.8/site-packages/astropy
[  100s] plugins: filter-subpackage-0.0.0, astropy-header-0.1.2, hypothesis-5.19.0, openfiles-0.5.0, mpl-0.11, remotedata-0.3.2, asdf-2.6.0, cov-2.10.0, doctestplus-0.7.0, arraydiff-0.3
[  155s] collecting ... collected 15020 items / 1 skipped / 15019 selected
...
[ 1012s] =================================== FAILURES ===================================
[ 1012s] _______________________ test_compound_fitting_with_units _______________________
[ 1012s] 
[ 1012s]     @pytest.mark.skipif('not HAS_SCIPY')
[ 1012s]     def test_compound_fitting_with_units():
[ 1012s]         x = np.linspace(-5, 5, 15) * u.Angstrom
[ 1012s]         y = np.linspace(-5, 5, 15) * u.Angstrom
[ 1012s]     
[ 1012s]         fitter = fitting.LevMarLSQFitter()
[ 1012s]         m = models.Gaussian2D(10*u.Hz,
[ 1012s]                               3*u.Angstrom, 4*u.Angstrom,
[ 1012s]                               1*u.Angstrom, 2*u.Angstrom)
[ 1012s]         p = models.Planar2D(3*u.Hz/u.Angstrom, 4*u.Hz/u.Angstrom, 1*u.Hz)
[ 1012s]         model = m + p
[ 1012s]     
[ 1012s]         z = model(x, y)
[ 1012s]         res = fitter(model, x, y, z)
[ 1012s]         assert isinstance(res(x, y), np.ndarray)
[ 1012s]         assert all([res[i]._has_units for i in range(2)])
[ 1012s]     
[ 1012s]         model = models.Gaussian2D() + models.Planar2D()
[ 1012s]         res = fitter(model, x, y, z)
[ 1012s]         assert isinstance(res(x, y), np.ndarray)
[ 1012s]         assert all([res[i]._has_units for i in range(2)])
[ 1012s]     
[ 1012s]         # A case of a mixture of models with and without units
[ 1012s]         model = models.BlackBody(temperature=3000 * u.K) * models.Const1D(amplitude=1.0)
[ 1012s]         x = np.linspace(1, 3, 10000) * u.micron
[ 1012s]     
[ 1012s]         with NumpyRNGContext(12345):
[ 1012s]             n = np.random.normal(3)
[ 1012s]     
[ 1012s]         y = model(x)
[ 1012s]         res = fitter(model, x, y * (1 + n))
[ 1012s]         # The large rtol here is due to different results on linux and macosx, likely
[ 1012s]         # the model is ill-conditioned.
[ 1012s] >       np.testing.assert_allclose(res.parameters, [3000, 2.1433621e+00, 2.647347e+00], rtol=0.4)
[ 1012s] E       AssertionError: 
[ 1012s] E       Not equal to tolerance rtol=0.4, atol=0
[ 1012s] E       
[ 1012s] E       Mismatched elements: 2 / 3 (66.7%)
[ 1012s] E       Max absolute difference: 1.2675296
[ 1012s] E       Max relative difference: 0.54769477
[ 1012s] E        x: array([3.000000e+03, 9.694539e-01, 3.914877e+00])
[ 1012s] E        y: array([3.000000e+03, 2.143362e+00, 2.647347e+00])
[ 1012s] 
[ 1012s] /usr/lib/python3.8/site-packages/astropy/modeling/tests/test_quantities_fitting.py:256: AssertionError
...
[ 1012s] = 1 failed, 14592 passed, 360 skipped, 68 xfailed, 65 warnings in 912.96 seconds =
```

(I am still waiting on the armv7l results while aarch64 passes)

### Steps to Reproduce
- Install astropy 4.1rc2 on a Linux with 32-bit libraries and run the unit tests

OR

- branch https://build.opensuse.org/package/show/home:bnavigator:astropy-rc/python-astropy-rc and play with it

### System Details

See above pytest header

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to remove 32-bit from testing and from generating wheels for releases. And since we don't support it, we might also be able to remove some code?

I cannot tell if I have overdone it or not, please review carefully.

There are also some stuff I don't dare to touch (e.g., C code, FITS). You can see them if you do a grep on "32bit", "32-bit", and "32 bit".

Close #10675 

### TODO

- [ ] Fix CI failures.
- [ ] Check "parallel and double" log carefully.

### CODE REVIEW

File: astropy/stats/tests/test_biweight.py [hunk 1]
```diff
@@ -558,15 +558,6 @@ def test_biweight_midcorrelation_inputs():
         biweight_midcorrelation(a2, a3)
 
 
-def test_biweight_32bit_runtime_warnings():
-    """Regression test for #6905."""
```
Comment (pllim) [inline astropy/stats/tests/test_biweight.py:562]: xref #6905

**Comment by pllim** [comment]
pre-commit.ci autofix

**Comment by olebole** [comment]
In Debian, we still have a number of official 32-bit architectures (ARM, i386, MIPS). Unless we run into massive test failures, I would like to keep Astropy available there; however this could probably be done as "32-bit is exotic now; don't put too much work into it".
However, I would ask to not remove 32-bit specific production code unless there is a specific reason for it. We should not artificially break 32-bit compatibility.
(not sure if this PR is the right place for this discussion; I probably missed the general one)

**Comment by taldcroft** [comment]
> However, I would ask to not remove 32-bit specific production code unless there is a specific reason for it.

In general astropy developers are free (and encouraged) to simplify the code base when possible. We do this frequently with old code related to working around bugs in dependencies -- as we stop supporting the older version then we take out that workaround code. I think the same should apply to 32-bit workarounds. Even if @pllim doesn't specifically find them here, eventually that code will get excised.

After we remove tests that are specific to 32-bit platforms and CI 32-bit testing, then I'm not sure Debian should be even making 32-bit packages for the newer astropy releases. I would have guessed that a packager like Debian would adhere to stated package requirements like numpy >= 22.0 and now bits == 64.

**Comment by saimn** [comment]
We should probably talk only about removing i686, not all 32bits distributions ? Because win32 is still probably used a lot. I didn't find numbers so just a guess.

**Comment by saimn** [comment]
Numpy stopped distributing i686 wheels (in 1.22 I think ? I could'nt find a reference) but they still distribute win32 wheels.
@olebole - How do you deal with Numpy for i386/i686 ?

File: astropy/stats/tests/test_biweight.py (hunk 1 above)
Comment (saimn) [reply astropy/stats/tests/test_biweight.py:562]: This should stay, `np.float32` arrays can be used on 64bit systems.

**Comment by bsipocz** [comment]
A lot of small telescopes, etc control software was running on raspberry's and 32bit, so I would keep those users in mind before purging the 32bit parts just for the sake of spring cleaning.

**Comment by pllim** [comment]
I think the main motivation is that numpy doesn't have 32bit anymore and it is a required dependency. So, are you saying that I should hold off on this until our numpy minversion is 1.22? Our minversion is 1.21 now.

https://github.com/astropy/astropy/blob/36e30a75d3ca704991d6d60d264476022d5b488d/setup.cfg#L31

Thanks for all the feedback but all the opinions here differ and now I am confused on what the actual ask was. @astrofrog said it is okay to drop in `main`.

**Comment by bsipocz** [comment]
> I think the main motivation is that numpy doesn't have 32bit anymore and it is a required dependency.

Not sure where this is sourced, numpy still builds 32bit windows wheels, and 32bit bits are still in the CI suite.

**Comment by pllim** [comment]
Yeah only win32 (https://pypi.org/project/numpy/#files). Does this mean we drop i686 but keep win32, as suggested above?

**Comment by bsipocz** [comment]
Just as Ole above, I'm mainly advocating for is to not artificially remove 32bit support from the code, whether it's tested with win32 or also with i686, I have no strong opinion about it.

**Comment by saimn** [comment]
So it seems Numpy still has a Linux 32bit CI job, e.g. https://github.com/numpy/numpy/pull/23360/checks?check_run_id=11868986979
Maybe we just remove the i686 wheels ? Better to keep a Linux job with 32bit if Debian still provides 32bit packages.

**Comment by pllim** [comment]
See https://github.com/astropy/astropy/pull/14517

**Pull request closed without merging**